File: Clients/Node/Admin/src/services/DatabaseBackupService.ts
import { BaseApiClient } from '../client/BaseApiClient';
import { ApiClientConfig } from '../client/types';
import type {
  BackupInfo,
  BackupResult,
  RestoreResult,
  BackupOptions,
  RestoreOptions,
  BackupValidationResult,
  BackupStorageStats,
  BackupDownloadResponse,
  BackupFilters,
  BackupProgress,
  BackupSystemStatus
} from '../models/databaseBackup';

/**
 * Service for managing database backups and restoration through the Admin API
 */
export class DatabaseBackupService extends BaseApiClient {
  private readonly baseEndpoint = '/api/database';

  constructor(config: ApiClientConfig) {
    super(config);
  }

  /**
   * Creates a new database backup
   * 
   * @param options - Optional backup configuration options
   * @returns Promise<BackupResult> Information about the created backup
   */
  async createBackup(options?: BackupOptions): Promise<BackupResult> {
    try {
      const backupInfo = await this.post<BackupInfo>(
        `${this.baseEndpoint}/backup`,
        options
      );

      return {
        success: true,
        backupInfo
      };
    } catch (error) {
      return {
        success: false,
        errorMessage: error instanceof Error ? error.message : 'Unknown error occurred'
      };
    }
  }

  /**
   * Retrieves a list of all available database backups
   * 
   * @returns Promise<BackupInfo[]> A list of backup information
   */
  async getBackups(): Promise<BackupInfo[]> {
    return this.get<BackupInfo[]>(`${this.baseEndpoint}/backups`);
  }

  /**
   * Gets information about a specific backup
   * 
   * @param backupId - The ID of the backup to retrieve
   * @returns Promise<BackupInfo | null> Backup information if found, null otherwise
   */
  async getBackupInfo(backupId: string): Promise<BackupInfo | null> {
    if (!backupId?.trim()) {
      throw new Error('Backup ID cannot be null or empty');
    }

    try {
      const backups = await this.getBackups();
      return backups.find(b => b.id === backupId) || null;
    } catch {
      return null;
    }
  }

  /**
   * Restores the database from a specific backup
   * 
   * @param backupId - The ID of the backup to restore from
   * @param options - Optional restore configuration options
   * @returns Promise<RestoreResult> The result of the restore operation
   */
  async restoreBackup(backupId: string, options?: RestoreOptions): Promise<RestoreResult> {
    if (!backupId?.trim()) {
      throw new Error('Backup ID cannot be null or empty');
    }

    try {
      await this.post(
        `${this.baseEndpoint}/restore/${encodeURIComponent(backupId)}`,
        options
      );

      return {
        success: true
      };
    } catch (error) {
      return {
        success: false,
        errorMessage: error instanceof Error ? error.message : 'Unknown error occurred'
      };
    }
  }

  /**
   * Downloads a backup file as ArrayBuffer
   * 
   * @param backupId - The ID of the backup to download
   * @returns Promise<BackupDownloadResponse> The backup file content and metadata
   */
  async downloadBackup(backupId: string): Promise<BackupDownloadResponse> {
    if (!backupId?.trim()) {
      throw new Error('Backup ID cannot be null or empty');
    }

    const response = await this.axios.get(
      `${this.baseEndpoint}/download/${encodeURIComponent(backupId)}`,
      {
        responseType: 'arraybuffer'
      }
    );

    const contentType = response.headers['content-type'] || 'application/octet-stream';
    const contentDisposition = response.headers['content-disposition'] || '';
    const fileNameMatch = contentDisposition.match(/filename="?([^"]+)"?/);
    const fileName = fileNameMatch ? fileNameMatch[1] : `backup-${backupId}.sql`;

    return {
      data: response.data,
      contentType,
      fileName,
      sizeBytes: response.data.byteLength
    };
  }

  /**
   * Downloads a backup file and returns it as a Blob (for browser environments)
   * 
   * @param backupId - The ID of the backup to download
   * @returns Promise<Blob> The backup file as a Blob
   */
  async downloadBackupAsBlob(backupId: string): Promise<Blob> {
    const downloadResponse = await this.downloadBackup(backupId);
    return new Blob([downloadResponse.data], { type: downloadResponse.contentType });
  }

  /**
   * Checks if a backup with the specified ID exists
   * 
   * @param backupId - The ID of the backup to check
   * @returns Promise<boolean> True if the backup exists, false otherwise
   */
  async backupExists(backupId: string): Promise<boolean> {
    if (!backupId?.trim()) {
      return false;
    }

    try {
      const backup = await this.getBackupInfo(backupId);
      return backup !== null;
    } catch {
      return false;
    }
  }

  /**
   * Gets the most recent backup
   * 
   * @returns Promise<BackupInfo | null> The most recent backup, or null if no backups exist
   */
  async getMostRecentBackup(): Promise<BackupInfo | null> {
    try {
      const backups = await this.getBackups();
      return backups
        .sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime())[0] || null;
    } catch {
      return null;
    }
  }

  /**
   * Gets backups created within a specific date range
   * 
   * @param startDate - The start date (inclusive)
   * @param endDate - The end date (inclusive)
   * @returns Promise<BackupInfo[]> Backups created within the specified date range
   */
  async getBackupsByDateRange(startDate: Date, endDate: Date): Promise<BackupInfo[]> {
    if (startDate > endDate) {
      throw new Error('Start date cannot be greater than end date');
    }

    const backups = await this.getBackups();
    return backups
      .filter(b => {
        const backupDate = new Date(b.createdAt);
        return backupDate >= startDate && backupDate <= endDate;
      })
      .sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());
  }

  /**
   * Gets backup storage statistics
   * 
   * @returns Promise<BackupStorageStats> Storage statistics for all backups
   */
  async getBackupStorageStats(): Promise<BackupStorageStats> {
    const backups = await this.getBackups();
    
    const totalBackups = backups.length;
    const totalSizeBytes = backups.reduce((sum, backup) => sum + backup.sizeBytes, 0);
    const averageSizeBytes = totalBackups > 0 ? Math.round(totalSizeBytes / totalBackups) : 0;

    const sortedByDate = backups.sort((a, b) => 
      new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime()
    );
    
    const oldestBackup = sortedByDate[0]?.createdAt;
    const newestBackup = sortedByDate[sortedByDate.length - 1]?.createdAt;
    
    const largestBackup = backups.sort((a, b) => b.sizeBytes - a.sizeBytes)[0];

    return {
      totalBackups,
      totalSizeBytes,
      totalSizeFormatted: this.formatBytes(totalSizeBytes),
      averageSizeBytes,
      averageSizeFormatted: this.formatBytes(averageSizeBytes),
      oldestBackup: oldestBackup ? new Date(oldestBackup) : undefined,
      newestBackup: newestBackup ? new Date(newestBackup) : undefined,
      largestBackup
    };
  }

  /**
   * Creates a backup and immediately downloads it
   * 
   * @param options - Optional backup configuration options
   * @returns Promise<BackupDownloadResponse> The backup file content and metadata
   */
  async createAndDownloadBackup(options?: BackupOptions): Promise<BackupDownloadResponse> {
    const backupResult = await this.createBackup(options);
    if (!backupResult.success || !backupResult.backupInfo) {
      throw new Error(`Failed to create backup: ${backupResult.errorMessage}`);
    }

    return this.downloadBackup(backupResult.backupInfo.id);
  }

  /**
   * Validates a backup file
   * 
   * @param backupId - The ID of the backup to validate
   * @returns Promise<BackupValidationResult> Validation results
   */
  async validateBackup(backupId: string): Promise<BackupValidationResult> {
    if (!backupId?.trim()) {
      throw new Error('Backup ID cannot be null or empty');
    }

    try {
      return this.post<BackupValidationResult>(
        `${this.baseEndpoint}/validate/${encodeURIComponent(backupId)}`
      );
    } catch (error) {
      return {
        isValid: false,
        errors: [error instanceof Error ? error.message : 'Validation failed'],
        warnings: []
      };
    }
  }

  /**
   * Deletes a backup file
   * 
   * @param backupId - The ID of the backup to delete
   */
  async deleteBackup(backupId: string): Promise<void> {
    if (!backupId?.trim()) {
      throw new Error('Backup ID cannot be null or empty');
    }

    await this.delete(`${this.baseEndpoint}/backups/${encodeURIComponent(backupId)}`);
  }

  /**
   * Gets filtered backups based on the provided criteria
   * 
   * @param filters - The filters to apply
   * @returns Promise<BackupInfo[]> Filtered list of backups
   */
  async getFilteredBackups(filters: BackupFilters): Promise<BackupInfo[]> {
    let backups = await this.getBackups();

    // Apply date filters
    if (filters.startDate) {
      backups = backups.filter(b => new Date(b.createdAt) >= filters.startDate!);
    }

    if (filters.endDate) {
      backups = backups.filter(b => new Date(b.createdAt) <= filters.endDate!);
    }

    // Apply size filters
    if (filters.minSizeBytes !== undefined) {
      backups = backups.filter(b => b.sizeBytes >= filters.minSizeBytes!);
    }

    if (filters.maxSizeBytes !== undefined) {
      backups = backups.filter(b => b.sizeBytes <= filters.maxSizeBytes!);
    }

    // Apply sorting
    if (filters.sortBy) {
      const sortDirection = filters.sortDirection === 'asc' ? 1 : -1;
      backups.sort((a, b) => {
        let aValue: any;
        let bValue: any;

        switch (filters.sortBy) {
          case 'createdAt':
            aValue = new Date(a.createdAt).getTime();
            bValue = new Date(b.createdAt).getTime();
            break;
          case 'sizeBytes':
            aValue = a.sizeBytes;
            bValue = b.sizeBytes;
            break;
          case 'fileName':
            aValue = a.fileName.toLowerCase();
            bValue = b.fileName.toLowerCase();
            break;
          default:
            return 0;
        }

        if (aValue < bValue) return -1 * sortDirection;
        if (aValue > bValue) return 1 * sortDirection;
        return 0;
      });
    } else {
      // Default sort by creation date (newest first)
      backups.sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());
    }

    // Apply pagination
    if (filters.page !== undefined && filters.pageSize !== undefined) {
      const startIndex = (filters.page - 1) * filters.pageSize;
      const endIndex = startIndex + filters.pageSize;
      backups = backups.slice(startIndex, endIndex);
    }

    return backups;
  }

  /**
   * Gets the current backup system status
   * 
   * @returns Promise<BackupSystemStatus> Current backup system status
   */
  async getBackupSystemStatus(): Promise<BackupSystemStatus> {
    try {
      return this.get<BackupSystemStatus>(`${this.baseEndpoint}/status`);
    } catch {
      // Fallback to basic status if endpoint is not available
      const backups = await this.getBackups();
      const mostRecent = await this.getMostRecentBackup();
      const stats = await this.getBackupStorageStats();

      return {
        isOperational: true,
        lastBackupDate: mostRecent?.createdAt ? new Date(mostRecent.createdAt) : undefined,
        availableBackups: backups.length,
        totalStorageUsed: stats.totalSizeFormatted,
        healthStatus: 'healthy',
        statusMessages: []
      };
    }
  }

  /**
   * Gets backup operation progress (if supported)
   * 
   * @param operationId - The operation ID to track
   * @returns Promise<BackupProgress | null> Progress information or null if not found
   */
  async getBackupProgress(operationId: string): Promise<BackupProgress | null> {
    if (!operationId?.trim()) {
      throw new Error('Operation ID cannot be null or empty');
    }

    try {
      return this.get<BackupProgress>(
        `${this.baseEndpoint}/progress/${encodeURIComponent(operationId)}`
      );
    } catch {
      return null;
    }
  }

  /**
   * Formats bytes into a human-readable string
   * 
   * @param bytes - The number of bytes
   * @returns A formatted string representation of the size
   */
  private formatBytes(bytes: number): string {
    const suffixes = ['B', 'KB', 'MB', 'GB', 'TB'];
    let counter = 0;
    let number = bytes;
    
    while (Math.round(number / 1024) >= 1 && counter < suffixes.length - 1) {
      number /= 1024;
      counter++;
    }
    
    return `${number.toFixed(1)} ${suffixes[counter]}`;
  }

  /**
   * Creates multiple backups in sequence
   * 
   * @param count - Number of backups to create
   * @param options - Optional backup configuration options
   * @returns Promise<BackupResult[]> Results of all backup operations
   */
  async createMultipleBackups(count: number, options?: BackupOptions): Promise<BackupResult[]> {
    if (count <= 0) {
      throw new Error('Count must be greater than 0');
    }

    const results: BackupResult[] = [];
    
    for (let i = 0; i < count; i++) {
      try {
        const result = await this.createBackup(options);
        results.push(result);
        
        // Add a small delay between backups to avoid overwhelming the system
        if (i < count - 1) {
          await new Promise(resolve => setTimeout(resolve, 1000));
        }
      } catch (error) {
        results.push({
          success: false,
          errorMessage: error instanceof Error ? error.message : 'Unknown error occurred'
        });
      }
    }

    return results;
  }

  /**
   * Deletes multiple backups
   * 
   * @param backupIds - Array of backup IDs to delete
   * @returns Promise<{ successCount: number; failedIds: string[]; errors: string[] }> Bulk operation results
   */
  async deleteMultipleBackups(backupIds: string[]): Promise<{
    successCount: number;
    failedIds: string[];
    errors: string[];
  }> {
    if (!backupIds || backupIds.length === 0) {
      return {
        successCount: 0,
        failedIds: [],
        errors: []
      };
    }

    let successCount = 0;
    const failedIds: string[] = [];
    const errors: string[] = [];

    for (const backupId of backupIds) {
      try {
        await this.deleteBackup(backupId);
        successCount++;
      } catch (error) {
        failedIds.push(backupId);
        errors.push(`Failed to delete backup ${backupId}: ${error instanceof Error ? error.message : 'Unknown error'}`);
      }
    }

    return {
      successCount,
      failedIds,
      errors
    };
  }
}

File: archive/Clients.DotNet/ConduitLLM.AdminClient/src/Services/DatabaseBackupService.cs
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Caching.Memory;
using ConduitLLM.AdminClient.Client;
using ConduitLLM.AdminClient.Models;

namespace ConduitLLM.AdminClient.Services;

/// <summary>
/// Service for managing database backups and restoration through the Admin API.
/// </summary>
public class DatabaseBackupService : BaseApiClient
{
    private const string BaseEndpoint = "/api/database";

    /// <summary>
    /// Initializes a new instance of the DatabaseBackupService class.
    /// </summary>
    /// <param name="httpClient">The HTTP client to use for requests.</param>
    /// <param name="configuration">The client configuration.</param>
    /// <param name="logger">Optional logger instance.</param>
    /// <param name="cache">Optional memory cache instance.</param>
    public DatabaseBackupService(
        HttpClient httpClient,
        ConduitAdminClientConfiguration configuration,
        ILogger<DatabaseBackupService>? logger = null,
        IMemoryCache? cache = null)
        : base(httpClient, configuration, logger, cache)
    {
    }

    #region Backup Operations

    /// <summary>
    /// Creates a new database backup.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Information about the created backup.</returns>
    public async Task<BackupResult> CreateBackupAsync(CancellationToken cancellationToken = default)
    {
        _logger?.LogDebug("Creating database backup");

        try
        {
            var backupInfo = await PostAsync<BackupInfo>($"{BaseEndpoint}/backup", cancellationToken: cancellationToken);

            var result = new BackupResult
            {
                Success = true,
                BackupInfo = backupInfo
            };

            _logger?.LogInformation("Created database backup {BackupId} ({Size})", 
                backupInfo.Id, backupInfo.SizeFormatted);
            return result;
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to create database backup");
            return new BackupResult
            {
                Success = false,
                ErrorMessage = ex.Message
            };
        }
    }

    /// <summary>
    /// Retrieves a list of all available database backups.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>A list of backup information.</returns>
    public async Task<IEnumerable<BackupInfo>> GetBackupsAsync(CancellationToken cancellationToken = default)
    {
        _logger?.LogDebug("Getting database backups list");

        try
        {
            var backups = await GetAsync<IEnumerable<BackupInfo>>($"{BaseEndpoint}/backups", cancellationToken: cancellationToken);

            _logger?.LogDebug("Retrieved {Count} database backups", backups?.Count() ?? 0);
            return backups ?? Enumerable.Empty<BackupInfo>();
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to get database backups");
            throw;
        }
    }

    /// <summary>
    /// Gets information about a specific backup.
    /// </summary>
    /// <param name="backupId">The ID of the backup to retrieve.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Backup information if found, null otherwise.</returns>
    public async Task<BackupInfo?> GetBackupInfoAsync(string backupId, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(backupId))
            throw new ArgumentException("Backup ID cannot be null or empty", nameof(backupId));

        try
        {
            var backups = await GetBackupsAsync(cancellationToken);
            return backups.FirstOrDefault(b => b.Id == backupId);
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to get backup info for {BackupId}", backupId);
            return null;
        }
    }

    #endregion

    #region Restore Operations

    /// <summary>
    /// Restores the database from a specific backup.
    /// </summary>
    /// <param name="backupId">The ID of the backup to restore from.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The result of the restore operation.</returns>
    public async Task<RestoreResult> RestoreBackupAsync(string backupId, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(backupId))
            throw new ArgumentException("Backup ID cannot be null or empty", nameof(backupId));

        _logger?.LogDebug("Restoring database from backup {BackupId}", backupId);

        try
        {
            // Use PostAsync without return type since restore typically returns success message
            await PostAsync($"{BaseEndpoint}/restore/{Uri.EscapeDataString(backupId)}", cancellationToken: cancellationToken);

            var result = new RestoreResult
            {
                Success = true
            };

            _logger?.LogInformation("Successfully restored database from backup {BackupId}", backupId);
            return result;
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to restore database from backup {BackupId}", backupId);
            return new RestoreResult
            {
                Success = false,
                ErrorMessage = ex.Message
            };
        }
    }

    #endregion

    #region Download Operations

    /// <summary>
    /// Downloads a backup file as a byte array.
    /// </summary>
    /// <param name="backupId">The ID of the backup to download.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The backup file content as a byte array.</returns>
    public async Task<byte[]> DownloadBackupAsync(string backupId, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(backupId))
            throw new ArgumentException("Backup ID cannot be null or empty", nameof(backupId));

        _logger?.LogDebug("Downloading backup {BackupId}", backupId);

        try
        {
            var response = await HttpClient.GetAsync($"{BaseEndpoint}/download/{Uri.EscapeDataString(backupId)}", cancellationToken);
            response.EnsureSuccessStatusCode();

            var content = await response.Content.ReadAsByteArrayAsync(cancellationToken);

            _logger?.LogInformation("Downloaded backup {BackupId} ({Size} bytes)", backupId, content.Length);
            return content;
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to download backup {BackupId}", backupId);
            throw;
        }
    }

    /// <summary>
    /// Downloads a backup file and saves it to the specified path.
    /// </summary>
    /// <param name="backupId">The ID of the backup to download.</param>
    /// <param name="filePath">The local file path to save the backup to.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    public async Task DownloadBackupToFileAsync(string backupId, string filePath, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(backupId))
            throw new ArgumentException("Backup ID cannot be null or empty", nameof(backupId));
        if (string.IsNullOrWhiteSpace(filePath))
            throw new ArgumentException("File path cannot be null or empty", nameof(filePath));

        _logger?.LogDebug("Downloading backup {BackupId} to {FilePath}", backupId, filePath);

        try
        {
            var content = await DownloadBackupAsync(backupId, cancellationToken);
            await File.WriteAllBytesAsync(filePath, content, cancellationToken);

            _logger?.LogInformation("Downloaded backup {BackupId} to {FilePath}", backupId, filePath);
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to download backup {BackupId} to {FilePath}", backupId, filePath);
            throw;
        }
    }

    /// <summary>
    /// Downloads a backup file as a stream.
    /// </summary>
    /// <param name="backupId">The ID of the backup to download.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>A stream containing the backup file content.</returns>
    public async Task<Stream> DownloadBackupAsStreamAsync(string backupId, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(backupId))
            throw new ArgumentException("Backup ID cannot be null or empty", nameof(backupId));

        _logger?.LogDebug("Downloading backup {BackupId} as stream", backupId);

        try
        {
            var response = await HttpClient.GetAsync($"{BaseEndpoint}/download/{Uri.EscapeDataString(backupId)}", cancellationToken);
            response.EnsureSuccessStatusCode();

            var stream = await response.Content.ReadAsStreamAsync(cancellationToken);

            _logger?.LogDebug("Started streaming backup {BackupId}", backupId);
            return stream;
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to download backup {BackupId} as stream", backupId);
            throw;
        }
    }

    #endregion

    #region Utility Operations

    /// <summary>
    /// Checks if a backup with the specified ID exists.
    /// </summary>
    /// <param name="backupId">The ID of the backup to check.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>True if the backup exists, false otherwise.</returns>
    public async Task<bool> BackupExistsAsync(string backupId, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(backupId))
            return false;

        try
        {
            var backup = await GetBackupInfoAsync(backupId, cancellationToken);
            return backup != null;
        }
        catch
        {
            return false;
        }
    }

    /// <summary>
    /// Gets the most recent backup.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The most recent backup, or null if no backups exist.</returns>
    public async Task<BackupInfo?> GetMostRecentBackupAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            var backups = await GetBackupsAsync(cancellationToken);
            return backups.OrderByDescending(b => b.CreatedAt).FirstOrDefault();
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to get most recent backup");
            return null;
        }
    }

    /// <summary>
    /// Gets backups created within a specific date range.
    /// </summary>
    /// <param name="startDate">The start date (inclusive).</param>
    /// <param name="endDate">The end date (inclusive).</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Backups created within the specified date range.</returns>
    public async Task<IEnumerable<BackupInfo>> GetBackupsByDateRangeAsync(
        DateTime startDate,
        DateTime endDate,
        CancellationToken cancellationToken = default)
    {
        if (startDate > endDate)
            throw new ArgumentException("Start date cannot be greater than end date");

        try
        {
            var backups = await GetBackupsAsync(cancellationToken);
            return backups.Where(b => b.CreatedAt >= startDate && b.CreatedAt <= endDate)
                         .OrderByDescending(b => b.CreatedAt);
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to get backups by date range");
            throw;
        }
    }

    /// <summary>
    /// Gets backup storage statistics.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Storage statistics for all backups.</returns>
    public async Task<object> GetBackupStorageStatsAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            var backups = await GetBackupsAsync(cancellationToken);
            var backupsList = backups.ToList();

            var totalSize = backupsList.Sum(b => b.SizeBytes);
            var averageSize = backupsList.Any() ? totalSize / backupsList.Count : 0;

            return new
            {
                TotalBackups = backupsList.Count,
                TotalSizeBytes = totalSize,
                TotalSizeFormatted = FormatBytes(totalSize),
                AverageSizeBytes = averageSize,
                AverageSizeFormatted = FormatBytes(averageSize),
                OldestBackup = backupsList.OrderBy(b => b.CreatedAt).FirstOrDefault()?.CreatedAt,
                NewestBackup = backupsList.OrderByDescending(b => b.CreatedAt).FirstOrDefault()?.CreatedAt,
                LargestBackup = backupsList.OrderByDescending(b => b.SizeBytes).FirstOrDefault()
            };
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to get backup storage statistics");
            throw;
        }
    }

    /// <summary>
    /// Creates a backup and immediately downloads it.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The backup file content as a byte array.</returns>
    public async Task<byte[]> CreateAndDownloadBackupAsync(CancellationToken cancellationToken = default)
    {
        _logger?.LogDebug("Creating and downloading backup");

        try
        {
            var backupResult = await CreateBackupAsync(cancellationToken);
            if (!backupResult.Success || backupResult.BackupInfo == null)
            {
                throw new InvalidOperationException($"Failed to create backup: {backupResult.ErrorMessage}");
            }

            var content = await DownloadBackupAsync(backupResult.BackupInfo.Id, cancellationToken);

            _logger?.LogInformation("Created and downloaded backup {BackupId}", backupResult.BackupInfo.Id);
            return content;
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to create and download backup");
            throw;
        }
    }

    #endregion

    #region Private Helper Methods

    /// <summary>
    /// Formats bytes into a human-readable string.
    /// </summary>
    /// <param name="bytes">The number of bytes.</param>
    /// <returns>A formatted string representation of the size.</returns>
    private static string FormatBytes(long bytes)
    {
        string[] suffixes = { "B", "KB", "MB", "GB", "TB" };
        int counter = 0;
        decimal number = bytes;
        while (Math.Round(number / 1024) >= 1)
        {
            number /= 1024;
            counter++;
        }
        return $"{number:n1} {suffixes[counter]}";
    }

    #endregion
}